### PR TITLE
fix: implement permission-based menu hiding in admin UI

### LINKED ISSUE #3416: [BUG][UI]: Create Team button shown to users without teams.create permission
### 🐞 Bug Summary
User can not add any public or private teams

---

### 🧩 Affected Component
Select the area of the project impacted:

- [ ] `mcpgateway` - API
- [x] `mcpgateway` - UI (admin panel)
- [ ] `mcpgateway.wrapper` - stdio wrapper
- [ ] Federation or Transports
- [ ] CLI, Makefiles, or shell scripts
- [ ] Container setup (Docker/Podman/Compose)
- [ ] Other (explain below)

---

### 🔁 Steps to Reproduce

1. Go to https://ocp2.cloud.boomerangplatform.net/dev/mcp-gateway/service/gateway/admin/#teams as an user 
2. Click Create New Team
3. Add values to all mandatory fields
4. Click Create Team

---
### 🤔Results
The Create Team is not doing anything. 403 error in network
{"detail":"Insufficient permissions. Required: teams.create"}

Image

---

### 🤔 Expected Behavior
If the user role do not have permission to Create New Team than the button should be disabled or not shown.

---

### 📓 Logs / Error Output
Paste any relevant stack traces or logs here.
⚠️ **Do not paste secrets, credentials, or tokens.**

---

### 🧠 Environment Info
You can retrieve most of this from the `/version` endpoint.

| Key | Value |
|-----|-------|
| Version or commit | `e.g. v0.9.0 or main@a1b2c3d` |
| Runtime | `e.g. Python 3.11, Gunicorn` |
| Platform / OS | `e.g. Ubuntu 22.04, macOS` |
| Container | `e.g. Docker, Podman, none` |

---

### 🧩 Additional Context (optional)
Add any configuration details, flags, or related issues.

### LINKED ISSUE #3554: [BUG][UI]: Non-admin users see menu items they lack permissions for
### 🐞 Bug Summary
Non-admin users see menu items in the Admin UI that they don't have permission to access, resulting in 403 Forbidden errors when clicked. This creates a poor user experience and confusion about what functionality is available to them.

---

### 🧩 Affected Component

- [x] `mcpgateway` - UI (admin panel)
- [ ] `mcpgateway` - API (RBAC enforcement)
- [ ] `mcpgateway.wrapper` - stdio wrapper
- [ ] Federation or Transports
- [ ] CLI, Makefiles, or shell scripts
- [ ] Container setup (Docker/Podman/Compose)
- [ ] Other (explain below)

---

### 🔁 Steps to Reproduce

1. Create a user with `non admin` role.
2. Login to the Admin UI at `/admin` with this user
3. Observe the sidebar menu
4. Click on "Users", "Maintenance", "Logs", or "Export/Import" menu items
5. Receive 403 Forbidden error

**Expected visible sections for developer role:**
- Tools ✅
- Resources ✅
- Prompts ✅
- Servers ✅

**Incorrectly visible sections (should be hidden):**
- Users ❌ (requires `admin.users.read`)
- Maintenance ❌ (requires `admin.maintenance`)
- Logs ❌ (requires `admin.logs.read`)
- Export/Import ❌ (requires `admin.export`)
- Plugins ❌ (requires `admin.plugins.read`)
- Metrics ❌ (requires `admin.metrics.read`)

---

### 🤔 Expected Behavior

**Menu items should be dynamically hidden based on user's RBAC permissions:**

1. **For developers:** Only see Tools, Resources, Prompts, Servers, Gateways
2. **For viewers:** Only see read-only sections (Tools, Resources, Prompts - read-only)
3. **For team admins:** See team management sections + core sections
4. **For platform admins:** See all sections

**No 403 errors should occur from clicking visible menu items.**

---

### 📓 Current Behavior / Error Output

**Scenario 1: Developer clicks "Users" menu item**
```
HTTP 403 Forbidden
{
  "detail": "Access denied"
}
```

**Scenario 2: Viewer clicks "Maintenance" menu item**
```
HTTP 403 Forbidden
{
  "detail": "Access denied"
}
```

**Root Cause Analysis:**

The Admin UI (`mcpgateway/admin.py` - `admin_ui()` function around line 3105) currently:

1. ✅ Loads user permissions via `PermissionService`
2. ✅ Enforces permissions on backend endpoints via `@require_permission` decorators
3. ❌ **Does NOT hide menu sections based on permissions**
4. ❌ Only uses static `UI_HIDDEN_SECTIONS` config and `is_admin` flag

**Current template logic (admin.html):**
```html
{% if is_admin and 'users' not in hidden_sections %}
<a href="#users" id="tab-users">
  <span>👤</span>
  <span>Users</span>
</a>
{% endif %}
```

**Problem:** `is_admin` is a boolean flag, not granular permission checking. Non-admin users with some permissions still see admin-only sections.

---

### 🧠 Environment Info

| Key | Value |
|-----|-------|
| Version or commit | `v0.9.0+` (affects all versions with RBAC) |
| Runtime | `Python 3.11+, Gunicorn/Uvicorn` |
| Platform / OS | `All platforms` |
| Container | `Docker, Podman, or native` |
| Affected Files | `mcpgateway/admin.py`, `mcpgateway/templates/admin.html` |

---

### 🧩 Additional Context

**Related Components:**
- RBAC implementation: `mcpgateway/middleware/rbac.py`
- Permission service: `mcpgateway/services/permission_service.py`
- Token scoping: `mcpgateway/auth.py` (`normalize_token_teams()`)

**Security Note:**
- Backend endpoints are properly protected via `@require_permission` decorators
- This is a **UX bug**, not a security vulnerability
- Users cannot bypass permissions by accessing hidden endpoints directly
- The issue is purely about UI visibility vs. actual permissions

**Impact:**
- **Severity:** Medium (UX issue, not security)
- **Frequency:** Affects all non-admin users
- **User Experience:** Poor - users encounter 403 errors frequently
- **Workaround:** None - users must learn which sections they can access

**Proposed Solution:**

Implement dynamic permission-based menu hiding:

1. **Backend (mcpgateway/admin.py):**
   - Add section-to-permission mapping
   - Check user permissions against required permissions
   - Add sections without permission to `hidden_sections`

2. **Template (admin.html):**
   - Remove redundant `is_admin` checks
   - Rely on `hidden_sections` for all visibility logic

3. **Configuration:**
   - Support both static (`UI_HIDDEN_SECTIONS`) and dynamic (permission-based) hiding
   - Static config takes precedence (always hide)

**Example Section-Permission Mapping:**
```python
SECTION_PERMISSIONS = {
    "users": "admin.users.read",
    "maintenance": "admin.maintenance",
    "logs": "admin.logs.read",
    "export-import": "admin.export",
    "plugins": "admin.plugins.read",
    "metrics": "admin.metrics.read",
    "performance": "admin.performance.read",
    "observability": "admin.observability.read",
    "tools": "tools.read",
    "servers": "servers.read",
    "resources": "resources.read",
    "prompts": "prompts.read",
    "gateways": "gateways.read",
    "teams": "teams.read",
    "tokens": "tokens.read",
}
```

**Testing Requirements:**
- Unit tests for permission-based hiding logic
- Integration tests for UI rendering with different roles
- Manual testing with developer, viewer, team_admin, platform_admin roles
- CLI tool for testing menu visibility: `python -m mcpgateway.cli.test_menu_visibility --user <email>`

**Documentation Updates:**
- `docs/docs/manage/rbac.md` - Add menu visibility behavior
- `docs/docs/manage/admin-ui-customization.md` - Update visibility configuration

**Related Issues:**
- Root cause identified during RBAC implementation analysis
- Affects user onboarding and training (users don't know what they can access)
- Increases support burden (users report "broken" menu items)

**Acceptance Criteria:**
- [ ] Users only see menu items they have permission to access
- [ ] No 403 errors from clicking visible menu items
- [ ] Static `UI_HIDDEN_SECTIONS` config still works
- [ ] Team-scoped visibility works correctly
- [ ] CLI tool available for testing visibility
- [ ] All user roles tested (platform_admin, team_admin, developer, viewer)
- [ ] Documentation updated
- [ ] Zero regression in existing functionality

### LINKED ISSUE #3619: [BUG][UI]: LLM Settings page remains in loading state for non-admin users
### 🐞 Bug Summary

When logged in as a non-admin user. The LLM Settings page in all tabs (Providers.,Models, Test) is in the loading state.

Image

### 🧩 Affected Component
Select the area of the project impacted:

- [ ] `mcpgateway` - API
- [x] `mcpgateway` - UI (admin panel)
- [ ] `mcpgateway.wrapper` - stdio wrapper
- [ ] Federation or Transports
- [ ] CLI, Makefiles, or shell scripts
- [ ] Container setup (Docker/Podman/Compose)
- [ ] Other (explain below)

---

### 🔁 Steps to Reproduce

1. Log in as a non-admin user
2. In the team selector, choose All teams
3. Navigate to the LLM Settings page and go through all tabs (Providers, Models, Test)
4. Observe that the loading state is at each tab

---

### 🤔 Expected Behavior

The LLM Settings page should not be in the loading state. We should display a blank page or the "no permission" message if a user has no privilege to use the LLM settings.

### PR DESCRIPTION
**Pull request by vishu-bh**
# 🐛 Bug-fix PR

Closes #3554
Closes #3416
Closes #3619

## 📌 Summary

Implements permission-based hiding in the admin UI so users only see sections and create actions they are allowed to use.

This PR fixes two related UI/RBAC regressions:
- `#3554`: non-admin users could see menu sections that returned 403 when opened
- `#3416`: users without `teams.create` could still see the `Create New Team` button and hit a 403 on submit

**Key architectural improvement**: Added automatic validation of section-to-permission mapping against route decorators to prevent drift.

## 🔁 Reproduction Steps

**Issue #3554**
1. Log in as a non-admin user such as `developer` or `viewer`
2. Open the admin UI
3. Navigate to a tab that requires higher permissions (e.g: `#users`, `#plugins`)
4. Before this fix, the request returned `403 Forbidden`

**Issue #3416**
1. Log in as a user without `teams.create`
2. Open the `Teams` section in the admin UI
3. Observe the `Create New Team` button
4. Click it and submit the form
5. Before this fix, the request returned `403 Forbidden` because the action was visible without the required permission

## 🐞 Root Cause

The admin UI mixed static hiding, a few `is_admin` checks, and no granular permission checks for most navigation and create actions. That let users see controls they could not actually use, resulting in avoidable `403` responses.

## 💡 Fix Description

**Core implementation:**

1. **SECTION_PERMISSIONS mapping** (`mcpgateway/admin.py`):
   - Explicit mapping of UI sections to required RBAC permissions
   - Examples: `"plugins": "admin.plugins.read"`, `"tools": "tools.read"`

2. **Automatic validation** (`mcpgateway/admin.py` + `mcpgateway/main.py`):
   - `_extract_permission_from_route()` introspects `@require_permission` decorators
   - `validate_section_permissions()` runs at startup to verify mapping consistency
   - Logs warnings if mappings drift from actual route decorators

3. **Permission-based section hiding** (`mcpgateway/admin.py`):
   - `get_hidden_sections_for_user()` evaluates section visibility with `PermissionService`
   - Static config and dynamic RBAC hiding are merged server-side
   - Fail-closed behavior hides sections on permission check errors

4. **Permission-based action hiding** (`mcpgateway/admin.py` + `mcpgateway/templates/admin.html`):
   - `UI_ACTION_PERMISSIONS` maps create controls to RBAC permissions
   - `get_user_action_permissions()` evaluates action visibility for the authenticated user
   - The `Create New Team` button is now hidden when `teams.create` is missing, addressing `#3416`

5. **Template cleanup** (`mcpgateway/templates/admin.html`):
   - Removed redundant inline admin-only checks
   - Relies on server-side permission flags for sections and actions

## 🧪 Verification

| Check | Command | Status |
|---|---|---|
| Unit tests | `pytest tests/unit/mcpgateway/test_admin_rbac_ui.py -q` | ✅ Passed |
| Existing unit and integration coverage on branch | Included in PR branch history | ✅ Present |
| Playwright regression coverage | `tests/playwright/test_admin_menu_visibility.py` | ✅ Updated to assert `Create New Team` visibility |

## 📐 MCP Compliance (if relevant)
- [x] N/A - UI/RBAC visibility fix only

## ✅ Checklist
- [x] Code formatted
- [x] No secrets committed
- [x] DCO sign-off included in commits
- [x] Unit tests cover create-action permission mapping
- [x] Playwright coverage now includes `#3416` regression behavior
- [x] Fail-closed behavior preserved
- [x] Section-to-route validation preserved

### CODE REVIEW

File: mcpgateway/templates/admin.html [hunk 1]
```diff
@@ -6880,6 +6895,166 @@ <h2 class="text-2xl font-bold dark:text-gray-200">
           </div>
         </div>
 
+        <!-- Old table for reference (hidden) -->
```
Comment (marekdano) [inline mcpgateway/templates/admin.html]: Why are we adding this old table back? I was removed in the PR https://github.com/IBM/mcp-context-forge/pull/3370 to fix Playwright tests.

File: mcpgateway/main.py [hunk 2]
```diff
@@ -7917,6 +7917,12 @@ async def cleanup_import_statuses(max_age_hours: int = 24, user=Depends(get_curr
 if ADMIN_API_ENABLED:
     logger.info("Including admin_router - Admin API enabled")
     app.include_router(admin_router)  # Admin routes imported from admin.py
+
+    # Validate section-to-permission mapping consistency at startup
+    # First-Party
+    from mcpgateway.admin import validate_section_permissions
```
Comment (marekdano) [inline mcpgateway/main.py:10846]: Is it ok to add it to the top of the file where `from mcpgateway.admin...` is?

File: mcpgateway/templates/admin.html (hunk 1 above)
Comment (vishu-bh) [reply mcpgateway/templates/admin.html]: I think this happened in merge. I will remove it.

**Comment by vishu-bh** [comment]
Thanks for the Review @marekdano, 

For this comment - "Both get_hidden_sections_for_user and get_user_action_permissions call check_permission() in a loop — ~24 calls for sections + 9 for actions = ~33 individual permission checks per admin page load. If check_permission hits the DB each time, this could be slow. Consider whether PermissionService has batch-check capability, or whether the results are cached. Since this only applies to /admin page loads it's probably fine in practice."

I was concerned about it to, but in ```permission_service.check_permission```, caching is already happening. So ~33 permission checks result in only 1 DB query per admin page load.

**Comment by crivetimihai** [comment]
Thanks @vishu-bh — fixes #3554 (non-admin users seeing inaccessible menu items). Targeting 1.1.0.

File: mcpgateway/admin.py [hunk 3]
```diff
@@ -212,6 +212,157 @@ class GrpcServiceNameConflictError(GrpcServiceError):  # type: ignore
...
+        )
+
+        if is_test_env:
+            # Hard error in test/CI to fail fast
+            raise ValueError(error_msg)
+        # Warning only in production to avoid breaking deployments
+        LOGGER.warning(error_msg)
+        LOGGER.warning("This may indicate the mapping needs updating.")
+    else:
+        LOGGER.info(f"SECTION_PERMISSIONS validation passed: all {len(SECTION_PERMISSIONS)} sections match route decorators")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: The value of `len(SECTION_PERMISSIONS)` is 24, but only 15 sections are actually validated (the ones in `_SECTION_TO_ROUTE_PATH`). Sections on other routers (version-info, settings, llm-providers, llm-models, llm-api-info and roots) are skipped. 

The log message implies all 24 sections were checked, maybe edit the message to say something like "all N mapped sections verified" or mention the gap.

File: mcpgateway/admin.py [hunk 4]
```diff
@@ -212,6 +212,157 @@ class GrpcServiceNameConflictError(GrpcServiceError):  # type: ignore
...
+        ValueError: In test/CI environments when mismatches are found
+    """
+    mismatches = []
+
+    for section, expected_perm in SECTION_PERMISSIONS.items():
+        route_path = _SECTION_TO_ROUTE_PATH.get(section)
+
+        if route_path is None:
+            # Section has no associated route (e.g., overview)
+            continue
```
Comment (gcgoncalves) [inline mcpgateway/admin.py:335]: `overview` is in `_SECTION_TO_ROUTE_PATH` (line 274). Maybe edit the comment to say "Section's route located on a different sub-router, skipping validation"

File: tests/playwright/test_admin_menu_visibility.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+            try:
+                element = page.locator(selector)
+                if element.count() > 0 and element.is_visible():
+                    # Click the tab
+                    element.click()
+                    page.wait_for_load_state("domcontentloaded")
+
+                    # Check for 403 error in page content
+                    content = page.content()
+                    assert "403" not in content and "Forbidden" not in content, f"Section '{section}' returned 403 error"
```
Comment (gcgoncalves) [inline tests/playwright/test_admin_menu_visibility.py]: Checking for the string "403" in the page HTML is not ideal: version strings, status codes in other contexts, or other text could contain "403" as a substring. The check should look for the actual response status instead.

File: mcpgateway/admin.py [hunk 6]
```diff
@@ -299,6 +450,155 @@ def get_ui_visibility_config(request: Request) -> Dict[str, Any]:
...
+        if section in hidden:
+            continue
+
+        # Skip sections with no permission requirement (defensive check for future sections)
+        if required_permission is None:
+            continue
+
+        # Check if user has the required permission
+        try:
+            has_permission = await permission_service.check_permission(
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: This runs for each permission, performing a trip to the db. Could be consolidated befor the `for` section, with something like below:

```python
user_perms = await permission_service.get_all_permissions_for_user(user_email, token_teams)
```

Then, checked like this:

```python
for section, permission in SECTION_PERMISSIONS.items():
    (...)
    if permission not in user_perms:
        hidden.add(section)
```

Comment (vishu-bh) [reply mcpgateway/admin.py]: Fixed it!

File: tests/playwright/test_admin_menu_visibility.py (hunk 5 above)
Comment (vishu-bh) [reply tests/playwright/test_admin_menu_visibility.py]: Makes sense. Addressed

File: mcpgateway/admin.py (hunk 6 above)
Comment (gcgoncalves) [reply mcpgateway/admin.py]: Good solution with the fallback. 👍

**Comment by gcgoncalves** [comment]
Testing, I could not see the menu items mentioned in the description in the first place, but trying to access them via the URL redirect me to the first available tab. The observability tab is available, though:

Screenshot 2026-03-21 at 00 00 28

At least, trying to access it results in a redirect.

**Comment by crivetimihai** [comment]
## Review Summary

Rebased onto main (clean, no conflicts) and reviewed the full PR. Found and fixed several issues:

### Fixes applied (2 commits)

**Commit 1: `fix: use valid RBAC permissions and harden batch path for public-only tokens`**

- **Invalid permission strings**: `SECTION_PERMISSIONS` used `admin.version.read` and `admin.settings.read` — neither exists in the `Permissions` class in `db.py`. No role grants them, so these sections were always hidden for non-wildcard roles. Replaced both with `admin.system_config` to match route decorators.
- **Security: public-only token bypass in batch path**: `get_hidden_sections_for_user()` batch `get_user_permissions()` path lacked the `token_teams=[]` guard for `admin.*` permissions that `check_permission()` enforces. A public-only token could see admin menu sections. Added the same guard.
- **Missing test coverage**: Added 6 tests covering batch permission path, wildcard grant, fallback on exception, `None` permission skip, production warning path, and public-only token denial.

**Commit 2: `fix: separate gRPC visibility from A2A, fix roots and monitoring guards`**

- **gRPC visibility leak**: The `agents` section bundled A2A and gRPC under `a2a.read`, but gRPC routes require `admin.grpc`. Users with only `a2a.read` (developer, viewer, team_admin) could see gRPC sidebar links, panels, and trigger cross-team gRPC service preloads. Split `grpc-services` into its own section with `admin.grpc` permission. Promoted from alias to first-class hidable section.
- **Roots permission mismatch**: `roots.read` doesn't exist in `Permissions` class; all roots routes use `admin.system_config`. Fixed the mapping.
- **Monitoring sidebar regression**: Performance and observability sidebar links lacked `is_admin` guard but their panels required it — non-admin users saw clickable tabs leading to empty panels. Added `is_admin` guard to sidebar links.
- **Regression tests**: Added 4 tests validating all `SECTION_PERMISSIONS` map to defined `Permissions` class constants, gRPC/A2A separation, roots mapping, and `grpc-services` as first-class section.

### What's correct and well-designed in the original PR

- Fail-closed behavior on all error paths
- Platform admin bypass with unrestricted tokens
- Template cleanup removing redundant `is_admin` checks
- Startup validation via `validate_section_permissions()` cross-referencing route decorators
- `_required_permission` metadata via `setattr` on decorator wrapper
- Batch `get_user_permissions()` optimization reducing DB queries
- Comprehensive test coverage (56 unit tests + Playwright E2E)

### Test results

- 56 PR-specific unit tests: all passing
- 1262 related tests (admin + config + passthrough): all passing
